app/components/request/configs.js

```diff
@@ -16,63 +16,106 @@ export default Component.extend(CanTriggerBuild, TriggerBuild, {
 
   repo: reads('request.repo'),
-  rawConfigs: or('requestConfig.rawConfigs', 'request.uniqRawConfigs'),
-  messages: reads('requestConfig.messages'),
-  loading: reads('requestConfig.loading'),
+  repoId: reads('repo.id'),
+  rawConfigs: or('preview.rawConfigs', 'request.uniqRawConfigs'),
+  messages: reads('preview.messages'),
+  loading: reads('preview.loading'),
 
   refType: 'sha',
-  sha: truncated('originalSha', 10),
+  sha: reads('originalSha'),
   branch: reads('originalBranch'),
   message: reads('request.commit.message'),
   config: reads('request.apiConfig.config'),
   mergeMode: reads('originalMergeMode'), // TODO store and serve merge mode for api request configs
   defaultMergeMode: 'deep_merge_append',
 
-  originalSha: or('requestSha', 'repoDefaultBranchLastCommitSha'),
+  originalSha: truncated('requestOrDefaultBranchSha', 10),
   originalBranch: or('requestBranch', 'repoDefaultBranch'),
   originalMergeMode: or('request.mergeMode', 'defaultMergeMode'),
   requestBranch: reads('request.branchName'),
   requestSha: reads('request.commit.sha'),
+  requestOrDefaultBranchSha: or('requestSha', 'repoDefaultBranchLastCommitSha'),
   repoDefaultBranch: reads('repo.defaultBranch.name'),
   repoDefaultBranchLastCommitSha: reads('repo.defaultBranch.lastBuild.commit.sha'),
 
-  onTriggerBuild(e) {
+  ref: computed('refType', 'branch', 'sha', function () {
+    return this.get(this.refType);
+  }),
+
+  didInsertElement() {
+    this.load();
+  },
+
+  onTrigger(e) {
     e.toElement.blur();
-    if (this.status == 'closed') {
+    if (this.closed) {
       this.set('status', 'open');
-    } else if (['open', 'customize', 'preview'].includes(this.status)) {
+    } else {
       this.submit();
     }
   },
 
   onCustomize() {
-    if (this.status == 'customize') {
+    if (this.customizing) {
       this.set('status', 'open');
       this.reset();
     } else {
       this.set('status', 'customize');
     }
   },
 
+  onPreview() {
+    if (!this.previewing) {
+      this.set('status', 'preview');
+      if (!this.loaded) {
+        this.load();
+      }
+    } else if (this.customized) {
+      this.set('status', 'customize');
+    } else {
+      this.set('status', 'open');
+    }
+  },
+
   onCancel() {
     this.set('status', 'closed');
     this.reset();
   },
 
+  load(debounce) {
+    if (this.customizing || this.previewing) {
+      this.preview.loadConfigs.perform(this.repoId, this.data, debounce);
+      this.set('loaded', true);
+    }
+  },
+
+  data: computed('repo', 'ref', 'mergeMode', 'config', function () {
+    return {
+      repo: {
+        slug: this.repo.get('slug'),
+        private: this.repo.get('private'),
+        default_branch: this.repo.get('defaultBranch.name'),
+      },
+      ref: this.ref,
+      mode: this.mergeMode,
+      config: this.config,
+      data: {
+        branch: this.branch,
+        commit_message: this.message
+      },
+      type: 'api'
+    };
```

Comment:
Some Javascript-fu to make it a little bit cleaner:
```javascript
    const { repo: repoRecord, ref, config, branch, mergeMode: mode, message } = this;
    const { slug, private, defaultBranch } = repoRecord;
    const repo = { slug, private, default_branch: defaultBranch.name };
    const data = { branch, commit_message: message };
    const type = 'api';
    
    return { repo, ref, mode, config, data, type };
```
Completely optional, of course ;)

Reply:
woha! i'd love to use that. unfortunately there seems to be a syntax error that i just cannot make sense of:

```
   97 |   data: computed('repo', 'ref', 'mergeMode', 'config', function () {
   98 |     const { repo: repoRecord, ref, config, branch, mergeMode: mode, message } = this;
>  99 |     const { slug, 'private', defaultBranch } = repoRecord;
      |                            ^
  100 |     const repo = { slug, private, default_branch: defaultBranch.name };
  101 |     const data = { branch, commit_message: message };
  102 |     const type = 'api';
```

Reply:
I think on the `const { slug, 'private', defaultBranch } = repoRecord;` line, removing the quotes around `private` may do it

Reply:
ah, sorry ... no, unfortunately not. copied the wrong part of the output. i had tried that one (Artur's original suggestion) first, and then thought maybe `private` is keyword that needs quoting.

```   98 |   data: computed('repo', 'ref', 'mergeMode', 'config', 'message', function () {
   99 |     const { repo: repoRecord, ref, config, branch, mergeMode: mode, message } = this;
> 100 |     const { slug, private, defaultBranch } = repoRecord;
      |                   ^
  101 |     const repo = { slug, private, default_branch: defaultBranch.name };
  102 |     const data = { branch, commit_message: message };
  103 |     const type = 'api';
```

but also, this whole thing is going to be subject to change. for instance, `config` now also needs to get an empty string as a default value ... i wouldn't know how to change this on my own.

Reply:
oh yeah, I forgot that `private` is reserved. For which I know a possible solution, but with the other things you mention, and since this was an optional suggestion in the first place, maybe it's worth revisiting later